eval-0trace/parse-each.go

```diff
@@ -0,0 +1,209 @@
+	IP  string
+}
+
+type Result struct {
+	Timestamp  string
+	ClientIP   string
+	Traceroute []Hops
+	Lasthop    string
+	Lastping   RtItem
+	ClientPing RtItem
+}
+
+type AvgRTTcompare struct {
+	Timestamp  string
+	ClientIP   string
+	LastHop    float64
+	ClientHop  float64
+	Difference float64
+}
+
+const ICMPCount = 5
+const ICMPTimeout = time.Second * 10
+
+type RtItem struct {
+	IP        string
+	PktSent   int
+	PktRecv   int
+	PktLoss   float64
+	MinRtt    float64
+	AvgRtt    float64
+	MaxRtt    float64
+	StdDevRtt float64
+}
+
+// IcmpPing conducts the ICMP ping measurement, code from proxy-detection-server
+func IcmpPing(ip string) (RtItem, error) {
+	pinger, err := ping.NewPinger(ip)
+	if err != nil {
+		log.Println(err)
+		return RtItem{}, err
+	}
+	pinger.Count = ICMPCount
+	pinger.Timeout = ICMPTimeout
+	err = pinger.Run() // Blocks until finished.
+	if err != nil {
+		log.Println(err)
+		return RtItem{}, err
+	}
+	stat := pinger.Statistics()
+	icmp := RtItem{ip, stat.PacketsSent, stat.PacketsRecv, stat.PacketLoss, fmtTimeMs(stat.MinRtt), fmtTimeMs(stat.AvgRtt), fmtTimeMs(stat.MaxRtt), fmtTimeMs(stat.StdDevRtt)}
+	return icmp, nil
+
+}
+
+// calcAvgDifference calculates the difference between the AvgRtt of the two given RtItem objects
+func calcAvgDifference(last RtItem, client RtItem) (float64, bool) {
+	if last.AvgRtt == 0 || client.AvgRtt == 0 {
+		return 0, false
+	}
+	return client.AvgRtt - last.AvgRtt, true
+}
+
+func fmtTimeMs(value time.Duration) float64 {
+	return (float64(value) / float64(time.Millisecond))
+}
+
+// parseTracerouteFile takes the path of a 0trace output file as input
+// and parses and returns the traceroute hops, clientIP from the path, and the last hop obtained by 0trace
+func parseTracerouteFile(tracefile string) ([]Hops, string, string) {
+	clientparts := strings.Split(tracefile, ".txt")
+	ipPath := clientparts[0]
+	ipsplit := strings.Split(ipPath, "/")
+	clientIP := ipsplit[len(ipsplit)-1]
+
+	readfile, _ := os.Open(tracefile) // opens in read mode
+	content := bufio.NewScanner(readfile)
+	found := false
+
+	var result []Hops
+	var lastHop string
+
+	// read all lines into slice
+	allLines := make([]string, 0)
+	for content.Scan() {
+		allLines = append(allLines, content.Text())
+	}
+
+	for _, lines := range allLines {
+		if strings.Contains(lines, zeroTraceBeginString) {
+			found = true
+			continue
+		}
+		// skip all lines prior to zeroTraceBeginString
+		if !found {
+			continue
+		}
+		parts := strings.Split(lines, " ")
+		hop := parts[0]
+		hopInt, _ := strconv.Atoi(hop)
+		if hopInt == 0 || len(parts) == 1 {
+			continue
+		} else {
+			ipaddr := parts[1]
+			currHop := Hops{Hop: hopInt, IP: ipaddr}
+			lastHop = ipaddr
+			result = append(result, currHop)
+		}
+	}
+	return result, clientIP, lastHop
+}
+
+// runIcmpPingForEval runs an ICMP Ping measurement for the given two IP (strings) and returns the RtItem for each
+func runIcmpPingForEval(lastHop string, clientIP string) (RtItem, RtItem) {
```

Comment:
Shall we check if `lastHop` and `clientIP` are identical and if so, return a latency difference of 0?